### ISSUE TITLE
DnD between wayland and xwayland not working

### ISSUE THREAD

**Issue by Bluey26** [labwc/labwc#1912]
Hi, i use `smplayer` as video player, and it has a function that lets me load a subtitle file by dropping it into the video. Smplayer is a mpv-based video player, and seems to also have this drag and drop function for subtitles.

`mpv` seems to recognize the subtitles i drop into the video box in labwc session, but smplayer is unable (for some reason).

~~I was suggested to run `WAYLAND_DEBUG=1 smplayer` to see some kind of message that could help where could be the problem, but that didnt show anything for me~~ 

The program is running through **Xwayland** .I checked the "smplayer log" option inside the program : "View" > "Smplayer log"

After dragging and dropping a subtitle file into the video box, the following message can be seen (in smplayer logs):

```
BaseGui::dragEnterEvent
BaseGui::dropEvent
BaseGui::dropEvent: list of dirs: ()
BaseGui::dropEvent: list of files: ()
BaseGui::dropEvent: subtitle file: ""
```

This seems to indicate that the subtitle file path is not recognized (somehow), but i always loaded subtitles from this path in X11 and worked.

Just for testing purposes i logged into a` gnome-wayland` session and the drag and drop function seems to work. This could discard that smplayer would not support that function in wayland.

Unfortunately this program does not have an option to show verbosity in the console, so this is the only i could find. I can give more logs of other commands if its needed.

OS: `Archlinux`
Labwc package and version: `labwc-git 0.7.2.r0.gb41af54-1`

---

**Update:**

Since the program runs in Xwayland, the issue seems to be the `wayland -> x11` clipboard not working. This was verified by running a Xwayland file manager:

` GDK_BACKEND=x11 thunar`

Then, reducing both windows sizes to make them to be visible in the screen at the same time. Dragging the subtitle file from thunar(using xwayland) to smplayer, seems to load the subtitle file.

The opposite way seems to not be working either: `x11 -> wayland`

This was tested with the same `thunar` launch command and `mpv` which runs natively in wayland. The subtitle was not loaded into mpv.

Talking with @Consolatis in the IRC room, we realized that both `xwayland -> xwayland `and` wayland->wayland ` seems to be working, so a workaround for me would be using a X11 thunar window to load the subtitles in smplayer, and if i want to load the subtitles into mpv i should use a native wayland file manager.

The corresponding wlroots issues, related to this, where:

https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/837

https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/760

https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/841

**Comment by hhhorb** [comment]
Weird, I can't confirm. Ran GDK_BACKEND=x11 thunar in a terminal, D&Ded an mp4 into mpv, it played, did the same with an *.srt subtitle file, the subtitles rendered.

BTW, as a former smplayer user, celluloid is a much better mpv front-end.

**Comment by hhhorb** [comment]
^ Just realized that while running GDK_BACKEND=x11 thunar, the field label in the window switcher still reads xdg-shell?

**Comment by Consolatis** [comment]
> ^ Just realized that while running GDK_BACKEND=x11 thunar, the 'type' label in the window switcher still reads xdg-shell, so I don't know.

I ran into the same issue, if there is already a thunar instance open, calling thunar again (even with `GDK_BACKEND` set) it will just spawn a new window in the running instance.

**Comment by hhhorb** [comment]
@Consolatis, thanks. I had thunar --daemon in my autostart. :^D

Bug confirmed.

**Comment by johanmalm** [comment]
@hhhorb Good to see you here :smile:

**Comment by hhhorb** [comment]
@johanmalm , thanks! I setup sid with labwc last month, and have been working with @01micko on his labwc repository...

https://github.com/01micko/01micko.github.io/

So far I've packaged Hyprlang/Hyprlock (screen lock utility) and Hyrpicker (wlroots color picker)...

https://github.com/01micko/01micko.github.io/tree/main/debian/pool/main/h

**Comment by jlindgren90** [comment]
I am not sure that I understand what is going on with the (now old) wlroots MRs, with the one open and the other closed.

However, I tested drag-and-drop from wayland (thunar) -> xwayland (audacious) in sway 1.10 (which is also using wlroots 0.18.x) and it works there. So it seems like we should be able to get it working in labwc also. From what I can tell, the other direction (xwayland -> wayland) is not supported by wlroots at all.

**Comment by Consolatis** [comment]
> However, I tested drag-and-drop from wayland (thunar) -> xwayland (audacious) in sway 1.10 (which is also using wlroots 0.18.x) and it works there.

Try setting `focus_follows_mouse no` (and ensure thunar is actually running in wayland mode, if you start it once with `GDB_BACKEND=x11` it will keep using that as long as the dbus session is active, even if closing the last window and opening a new one).

> So it seems like we should be able to get it working in labwc also.

Actually it does (somewhat) if you enable followMouse in labwc.

> From what I can tell, the other direction (xwayland -> wayland) is not supported by wlroots at all.

Yes, I think you are right. From what I remember there exists an old MR for that somewhere but it isn't ready.

https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4928 should fix xdg->xwayland.

**Comment by jlindgren90** [comment]
Thanks @Consolatis - I will give your MR a try.

**Comment by jlindgren90** [comment]
> https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4928 should fix xdg->xwayland.

It does indeed, thanks! Tested with Thunar (xdg-shell) -> Audacious (xwayland).

**Comment by Consolatis** [comment]
Wayland -> Xwayland drag and drop should be working by using the just released wlroots 0.18.2. Please comment if that is not the case.

Xwayland -> Wayland drag and drop not working is known and due to a missing implementation in wlroots.
